CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+      if a_dir == "import":
+        discrim = values.assert_next(i32)
+        ptr = values.assert_next(i32)
+        len = values.assert_next(i32)
+        trap_if_out_of_bounds(discrim)      # must be 0 or 1
+        if discrim == 0:
+          handle = TABLE.register_buffer(type, a, ptr, len)
+        else:
+          handle = ptr
+          trap_if_out_of_bounds(len)        # must be 0
+      else:
+        handle = values.assert_next(i32)
+
+      # afterwards this would basically transfer the integer handle into the
+      # destination module the same way as `shared-handle`
+
+      # ...
+    }
+```
+
+Here the `adapt` function's goal is to do all the translation and `adapt` is the
+sole source of trusted truth in terms of validity of values in the canonical
+ABI. It's expected, though, that all interpretations of the canonical ABI will
+have `adapt` somewhere in them.
+
+# Intrinsics
+
+One of the final details of the canonical ABI is that there will be a few
+intrinsic functions to operate on the various types. These are primarily
+intended for handle-related types and lists.
+
+## Memory Intrinsics
+
+Lists sometimes need to be communicated between modules and this necessitates
+the need to allocate and free memory in modules. If needed the malloc/free
+functions will be exposed as:
+
+```wasm
+(module
+  (func (export "witx_malloc") (param i32 i32) (result i32)
+    ;; ...
+  )
+  (func (export "witx_free") (param i32 i32 i32)
+    ;; ...
+  )
+)
+```
+
+The `witx_malloc` function takes two arguments, a size and an alignment. It
+returns an `i32` which is a pointer in memory which has the desired alignment
+and is valid for `size` bytes. It's expected that the function will trap if
+allocation fails.
+
+The `witx_free` function takes a pointer/size/alignment and informs the
+allocator that the memory is no longer needed.
+
+## Handle Intrinsics
+
+For all types `handle $resource` imported and used by a module the module can
+also import a function called `${resource}_drop` (the name of the resource
+prepended with `_drop`). This destructor function is present to inform the
+management of tables that the module calling the drop function no longer needs
+the resource. This will remove the specified capability from the module. The
+function will trap if the specified handle is invalid.
+
+Modules exporting a resource `$resource` are similarly expected to export a
+`${resource}_drop` function. This function will automatically be called by the
+table management machinery when necessary, or notably when all references to the
+handle have been removed.
+
+This intends to imply that handles have reference counting associated with them.
+The reference count is how many wasm modules are currently referring to the
+handle. The reference count increases as the resource crosses module boundaries,
+and the reference count decreases when a module willingly gives up the handle to
+the resource (by calling the drop function).
+
```

Comment:
I think we need two more intrinsics so that wasm is able to virtualize interfaces with handles:
* an intrinsic for creating a new handle (implemented by the caller) in which the caller passes an `i32` (probably a pointer into linear memory)
* an intrinsic for getting back the `i32` for a handle, but only if that handle is `owned_by_current_module` (trapping otherwise)

Reply:
Oh this was intended to be covered automatically in serialization/deserialization/adaptation. Creating a handle is simply passing the i32 out to another module and re-acquiring the i32 happens when the handle is passed back to you automatically.

I'm not sure why other intrinsics would be needed though, so you may need to walk me through what you're thinking here

Reply:
Nah that all makes sense to me. I'll switch to an intrinsic like this.

Reply:
Hm ok actually thinking about this I'm less certain that this is a good idea. This I think muddies the ownership story and makes for some unnecessary reference counting.

For example if a module wants to return a handle to something it just created, it'd first call an intrinsic to wrap up a private 32-bit value into a handle. This creates a handle with reference count 1. Ideally the module would transfer ownership of that handle to the outside world, but that's not how handles are specified right now in lifting/lowering (that requires the unique/borrowed/etc). This means that the callee now has a strong reference count of 1 which it didn't have anywhere to store since it just returned it. It's not clear when the decrement should happen?

Similarly though if a module receives a handle from the outside world, what it ends up receiving is a fresh index from the outside world for a new handle given the existing lifting/lowering. It would then call an intrinsic to acquire the internal private 32-bit value on this handle. After the call finishes it would have to then deallocate this owned handle.

Overall I think what you're saying here makes sense, but I think it only makes sense in a world where there's shared/borrowed/unique handles. The current lift/lower for handles I've documented effectively assumes that returning handles is always returning a unique handle, and taking a handle always takes a borrowed handle. I think what you're spec'ing would be "what if both cases were shared handles".

Typing this all out I realize that handles are intended to be shared handles today, but this feels like a lot of overhead for something that probably won't be desired that much. There's overhead of extra calls in the "receive a handle" case, and in the "return a handle" case I'm not actually sure where the decrement of the reference count would generally go...

Reply:
In some sense, you've already implemented a shared handle, since sharing is what happens when A passes a handle to B, with neither A nor B being the owner.  What I'm proposing is that things work the same way even when A is the resource owner (b/c A will have already created the resource, sticking a Handle into A's own TABLE).  I *think* this answers the ref-counting questions?

You're right that there is unwanted extra builtin-call overhead, though.  One idea is to say that the canonical ABI for lifting a handle (when you are the resource type owner) is to pass two `i32`s: the first one says "create a new one, the second `i32` is the private value", the second one says "i already created the handle, the second `i32` is the index in my handle table".  That means that the canonical ABI for a handle type depends on the resource type, but I think that's fine, since it's static.  WDYT?